base/loading.jl

```diff
@@ -408,9 +408,33 @@ end
 ## generic project & manifest API ##
 
 const project_names = ("JuliaProject.toml", "Project.toml")
-const manifest_names = ("JuliaManifest.toml", "Manifest.toml")
+const _manifest_names = ("JuliaManifest.toml", "Manifest.toml")
 const preferences_names = ("JuliaLocalPreferences.toml", "LocalPreferences.toml")
 
+function manifest_names()
+    custom_def = get(ENV, "JULIA_MANIFEST_NAME", nothing)
+    if !isnothing(custom_def) && !isempty(custom_def)
+        if !endswith(custom_def, ".toml")
+            throw(ArgumentError("Manifest name provided by `JULIA_MANIFEST_NAME` must end with `.toml`. Provided: $(repr(custom_def))"))
+        end
+        return (populate_manifest_name(custom_def),)
+    else
+        return _manifest_names
+    end
+end
+
+function populate_manifest_name(name::String)
+    for vpart in (VERSION.major, VERSION.minor, VERSION.patch)
+        name = replace(name, "#" => "$vpart", count = 1)
+    end
```

Comment:
I would suggest that you factor this code out into a separate function, e.g.

```julia
function expand_version_placeholders(str::String)
    for vpart in (VERSION.major, VERSION.minor, VERSION.patch)
        str = replace(str, "#" => "$vpart", count = 1)
    end
    return str
end
```

Then, the `populate_manifest_name` function will call the `expand_version_placeholders` function. And also, the `load_path_expand` function will call the `expand_version_placeholders` function, thus reducing code duplication.

Specifically, inside the `load_path_expand` function, we would delete the following three lines and replace them with a call to `expand_version_placeholders`:

https://github.com/JuliaLang/julia/blob/f879d2092c19f342a838b1c084464be19f68e97e/base/initdefs.jl#L255-L257